relay-server/src/services/processor/nnswitch.rs

```diff
@@ -0,0 +1,428 @@
+fn expand_dying_message_from_envelope_items(data: Bytes, envelope: &mut Envelope) -> Result<()> {
+    let items =
+        Envelope::parse_items_bytes(data).map_err(SwitchProcessingError::EnvelopeParsing)?;
+    for item in items {
+        // If it's an event type, merge it with the main event one already in the envelope.
+        if item.ty() == &ItemType::Event {
+            if let Some(event) = envelope.get_item_by_mut(|it| it.ty() == &ItemType::Event) {
+                update_event(item, event).map_err(SwitchProcessingError::InvalidJson)?;
+                // Don't add this item as a new envelope item now that it's merged.
+                continue;
+            }
+        }
+        envelope.add_item(item);
+    }
+    Ok(())
+}
+
+fn update_event(item: Item, event: &mut Item) -> std::result::Result<(), serde_json::Error> {
+    let original_json = serde_json::from_slice::<serde_json::Value>(&event.payload())?;
+    let mut new_json = serde_json::from_slice(&item.payload())?;
+    utils::merge_values(&mut new_json, original_json);
+    let new_payload = serde_json::to_vec(&new_json)?;
+    event.set_payload(ContentType::Json, new_payload);
+    Ok(())
+}
+
+fn decompress_data(
+    payload: Bytes,
+    compressed_length: usize,
+    compression: u8,
+    compression_arg: u8,
+) -> Result<Bytes> {
+    if payload.len() < compressed_length {
+        return Err(SwitchProcessingError::InvalidValue(
+            "compressed data length".into(),
+            compressed_length,
+        ));
+    }
+
+    let data = payload.slice(0..compressed_length);
+    match compression {
+        // No compression
+        0 => Ok(data),
+        // Zstandard
+        1 => decompress_data_zstd(data, compression_arg)
+            .map(Bytes::from)
+            .map_err(SwitchProcessingError::Zstandard),
+        _ => Err(SwitchProcessingError::InvalidValue(
+            "compression format".into(),
+            compression as usize,
+        )),
+    }
+}
+
+static ZSTD_DICTIONARIES: &[&[u8]] = &[
+    // index 0 = empty dictionary (a.k.a "none")
+    b"",
+];
+
+static ZSTD_DEC_DICTIONARIES: OnceLock<[zstd::dict::DecoderDictionary; ZSTD_DICTIONARIES.len()]> =
+    OnceLock::new();
+fn decompress_data_zstd(data: Bytes, dictionary_id: u8) -> std::io::Result<Vec<u8>> {
+    // We initialize dictionaries only once and reuse them when decompressing.
+    let dictionaries = ZSTD_DEC_DICTIONARIES.get_or_init(|| {
+        let mut dictionaries: [zstd::dict::DecoderDictionary; ZSTD_DICTIONARIES.len()] =
+            [zstd::dict::DecoderDictionary::new(ZSTD_DICTIONARIES[0])];
+        for i in 0..ZSTD_DICTIONARIES.len() {
+            dictionaries[i] = zstd::dict::DecoderDictionary::new(ZSTD_DICTIONARIES[i]);
+        }
+        dictionaries
+    });
+
+    let dictionary = dictionaries
+        .get(dictionary_id as usize)
+        .ok_or(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "Unknown compression dictionary",
+        ))?;
```

Comment:
Maybe this would look better if this was in its own function `id -> Dictionary`, you can also inline the statics into that function body.